### PR TITLE
[Requirements resolver] Initial integration with the state machine v3

### PR DESCRIPTION
**Pull request by willianrampazzo**
**NOTES**:
- The intent of this PR is to handle the architecture of the requirements resolver. Specifically, the addition of a resolver, changes to the state machine, and changes to the task creation.
- To help with the architecture part, a package requirement runner was added. IMHO, according to the architecture presented here, the runner seems to be good enough to be considered fully implemented.
- There are **massive** changes that need to be made to the Human UI to display a nice output. I would like to handle that in a different PR.
- As the Human UI is not reflecting the reality of the tests, I would like this feature to be considered as **experimental**.
- If this version does not have architectural changes, in the next PR version I will add the documentation and unit/functional tests.

This is a minimum set of changes to introduce and integrate the Requirements Resolver with the state machine. The following tests enable the code of this PR.

A single test that passes (`single_success.py`):

```python
from avocado import Test


class PassTest(Test):

    """
    Example test that passes.

    :avocado: tags=fast
    :avocado: requirement={"type": "package", "name": "vim-common"}
    """

    def test(self):
        """
        A test doesn't have to fail to pass
        """
```

Expected output:

```
$ avocado run --test-runner=nrunner single_success.py 
JOB ID     : 9cc29be300f2620006b3290508dbd32127a1877d
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-9cc29be/job.log
 (1/1) single_success.py:PassTest.test: STARTED
 (1/1) single_success.py:PassTest.test: PASS (0.15 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-9cc29be/results.html
JOB TIME   : 7.12 s
```

A single test that fails (`single_fail.py`):

```python
from avocado import Test


class FailTest(Test):

    """
    :avocado: requirement={"type": "package", "name": "vim-common"}
    :avocado: requirement={"type": "package", "name": "bla"}
    :avocado: requirement={"type": "file", "name": "foo.bar"}
    """

    def test(self):
        """
        A fail test
        """
```

Expected output:

```
$ avocado run --test-runner=nrunner single_fail.py 
JOB ID     : 3a24d81bbb1d714461978e45fa6935a89bc14143
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/job.log
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 1 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/results.html
JOB TIME   : 6.17 s
```

**NOTE**: as the test task never starts, it is not accounted for in the Human UI, so it does not show in the execution or in the results.

Multiple tests that pass (`multipletests_success.py`):

```python
from avocado import Test


class PassTest(Test):

    """
    Example test that passes.

    """

    def test_a(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-common"}
        """
    def test_b(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-minimal"}
        """
    def test_c(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-common"}
        """
```

Expected output:

```
$ avocado run --test-runner=nrunner multipletests_success.py 
JOB ID     : 4e2b0a3217838c70a596a125c41ce26387d186fe
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.28-4e2b0a3/job.log
 (1/3) multipletests_success.py:PassTest.test_a: STARTED
 (1/3) multipletests_success.py:PassTest.test_a: PASS (0.14 s)
 (2/3) multipletests_success.py:PassTest.test_b: STARTED
 (2/3) multipletests_success.py:PassTest.test_b: PASS (0.14 s)
 (3/3) multipletests_success.py:PassTest.test_c: STARTED
 (3/3) multipletests_success.py:PassTest.test_c: PASS (0.14 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.28-4e2b0a3/results.html
JOB TIME   : 9.14 s
```

Multiple tests where two fail (`multipletests_fail.py`):

```python
from avocado import Test


class FailTest(Test):

    """
    Example test that fails.

    """

    def test_a(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-common"}
        :avocado: requirement={"type": "package", "name": "bla"}
        """
    def test_b(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-common"}
        """
    def test_c(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-minimal"}
        :avocado: requirement={"type": "package", "name": "bla"}
        """
```

Expected output:

```
$ avocado run --test-runner=nrunner multipletests_fail.py 
JOB ID     : c04c72ebbea91ce45f91cc36c50a5de64c114368
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.29-c04c72e/job.log
 (2/3) multipletests_fail.py:FailTest.test_b: STARTED
 (2/3) multipletests_fail.py:FailTest.test_b: PASS (0.14 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 2 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.29-c04c72e/results.html
JOB TIME   : 12.17 s
```

**NOTE**: again, as the test tasks never start, they are not accounted for in the Human UI, so it does not show them in the execution or in the results.

How to play with the requirements package runner:

```
$ avocado-runner-requirement-package runnable-run action=check name=vim-common
{'status': 'started', 'time': 104504.359667453}
{'status': 'running', 'time': 104504.861630231}
{'status': 'running', 'time': 104505.362315153}
{'result': 'pass', 'returncode': 0, 'stdout': b'', 'stderr': b"dnf module for Python is required. Using the basic support from rpm and dnf commands\nRunning 'rpm -q vim-common'\nCommand 'rpm -q vim-common' finished with 0 after 0.011124849319458008s\nPackage vim-common already installed\n", 'status': 'finished', 'time': 104505.362690156}
```

Changes from [v2](https://github.com/avocado-framework/avocado/pull/4481):
- Change the `avocado-runner-requirement-package` to handle kwargs instead of args.
- Add support to `check` and `remove` packages to the `avocado-runner-requirement-package`.
- Move the requirements resolver back to its own file at `avocado/core/requirements/resolver.py`.
- Add missing support to skip a test if one of its dependencies failed during triage or during the execution.

Changes from [v1](https://github.com/avocado-framework/avocado/pull/4423):
- Requirements are now handled as runners. Each requirement type is mapped to a specific runner. The addition of a new requirement type should be as simple as creating a new runner called `requirement_newtype.py`.
- Moved the resolver to the `avocado/plugins/resolvers.py` file as it is much simpler now.
- Changed the attribute name of the Task class to `dependency` instead of `prereqs`. This maps well to a dependency graph.
- Make the dependencies a set and improve the handling of it in the state machine.

Changes from [v0](https://github.com/avocado-framework/avocado/pull/4369):
- Improve docstring of RequirementRunner;
- Change the Requirements Resolver entry point to a static method;
- Variables renaming to make them more meaningful;
- Improve log message when the requirement type is not available;
- Classes reorder inside the file.

### CODE REVIEW

File: avocado/core/task/statemachine.py
Comment (beraldoleal) [inline avocado/core/task/statemachine.py:117]: nitpick: To improve readability I would suggest adding a `AND` here.

File: avocado/core/task/statemachine.py
Comment (beraldoleal) [inline avocado/core/task/statemachine.py:140]: This method is too big, do you think would be easy to split into small ones?

File: avocado/plugins/runner_nrunner.py
Comment (beraldoleal) [inline avocado/plugins/runner_nrunner.py:162]: nitpick: again, to improve readability, I would suggest invert the order here:

```python
if runnable.requirements is None:
   return runtime_tasks
...
```

**Review by beraldoleal — CHANGES_REQUESTED** [review]
Hi @willianrampazzo, I left a few comments for you.

Besides those comments, my only concern is about the output. I'm not sure how much you already stressed this with @clebergnu but IMO, it feels strange to have the same results as tests. requirements and tests should be different things.

For instance:

```
 (1/1) requirement-package-vim-common: STARTED
 (1/1) requirement-package-bla: STARTED
 (1/1) requirement-package-bla: ERROR (3.01 s)
 (1/1) requirement-package-vim-common: PASS (4.51 s)
RESULTS    : PASS 1 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```
In this example, we have *one* test with multiple requirements, so the sum of results should be 1. 1 Pass + 1 Error gives us the impression that we had 2 tests.

Unless I'm missing something, those results should be related to tests, anything else could be "Skip". Saying that a test has an error, could lead us to the impression that the error it is with the "test itself". I had understood that everything else (bootstrap and cleanup) would lead us to skip. And IMO, installing requirements are part of the bootstrap. 

Also, even with the "requirement-" prefix it is confusing adding those requirements to the list of tests, maybe just an "indentation" would be enough, or maybe we could suppress that from the output, and only show this with debug=true, not sure.

File: avocado/plugins/runner_nrunner.py
Comment (willianrampazzo) [reply avocado/plugins/runner_nrunner.py:162]: But I need to process all the runnables before I can return the `runtime_tasks`. The return is out of the `for`.

File: avocado/core/task/statemachine.py
Comment (willianrampazzo) [reply avocado/core/task/statemachine.py:117]: I did it in the first implementation, but the line exceeds the 80 column limit, so I thought it would be more readable, breaking it into 2 `if` statements. Anyway, I don't have a problem merging them and breaking the line.

Comment (willianrampazzo) [reply avocado/core/task/statemachine.py]: I'm merging the lines and sending the self-test fix in a force push.

File: avocado/core/task/statemachine.py
Comment (willianrampazzo) [reply avocado/core/task/statemachine.py:140]: The method is 64 lines long; removing the comments and blank lines makes it 45 lines. The blocks that update the state machine queue use 2 or 3 lines because we need to lock the state machine queue, and we have 6 occurrences of them. Personally, I don't see a need to split because most of the code is "accessory", but if it helps in readability, I can move lines 109-136 to a new function.

Comment (beraldoleal) [reply avocado/core/task/statemachine.py:140]: Yes, big is relative, but 64 lines are too big IMO. I try to follow the rule of 30". :) And I appreciate this small split. Thank you.

File: avocado/plugins/runner_nrunner.py
Comment (beraldoleal) [reply avocado/plugins/runner_nrunner.py:162]: Ohh I missed that, but still maybe you could do this:

```python
if runnable.requirements is None:
   continue
...
...
```

Comment (willianrampazzo) [reply avocado/plugins/runner_nrunner.py:162]: I see. The result is the same, but we reduce one level of indentation. I'll change that.

**Comment by willianrampazzo** [comment]
@clebergnu @beraldoleal @richtja I added the commits from PR #4543 here to make it easy to test this PR.

**Comment by willianrampazzo** [comment]
> Hi @willianrampazzo, I left a few comments for you.
> 
> Besides those comments, my only concern is about the output. I'm not sure how much you already stressed this with @clebergnu but IMO, it feels strange to have the same results as tests. requirements and tests should be different things.
> 
> For instance:
> 
> ```
>  (1/1) requirement-package-vim-common: STARTED
>  (1/1) requirement-package-bla: STARTED
>  (1/1) requirement-package-bla: ERROR (3.01 s)
>  (1/1) requirement-package-vim-common: PASS (4.51 s)
> RESULTS    : PASS 1 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
> ```
> 
> In this example, we have _one_ test with multiple requirements, so the sum of results should be 1. 1 Pass + 1 Error gives us the impression that we had 2 tests.
> 
> Unless I'm missing something, those results should be related to tests, anything else could be "Skip". Saying that a test has an error, could lead us to the impression that the error it is with the "test itself". I had understood that everything else (bootstrap and cleanup) would lead us to skip. And IMO, installing requirements are part of the bootstrap.
> 
> Also, even with the "requirement-" prefix it is confusing adding those requirements to the list of tests, maybe just an "indentation" would be enough, or maybe we could suppress that from the output, and only show this with debug=true, not sure.

I updated the _expected output_ to reflect the final results with PR #4543.

**Comment by PraveenPenguin** [comment]
wow i am really waiting for this , thanks @willianrampazzo , will review this as well test

**Comment by beraldoleal** [comment]
Hi @willianrampazzo, I do understand the motivations behind merging commits from the second PR here. But since it is there, I will consider this as a different PR and wait for the approval there (and a rebase here), before doing my second round here.

**Comment by willianrampazzo** [comment]
Code rebased to remove the commits related to the output changes.

**Comment by clebergnu** [comment]
@willianrampazzo I understand you did a rebase, but are you the Task category has been accounted for here?  This is the result of my experiments:

```
$ git rev-parse HEAD
e1b4bd5d8e85a9364f9dc9f2ad2f8f13b3b879c0

$ avocado run --test-runner=nrunner single_success.py 
JOB ID     : 9e8c9e71d250c40facd4039e529f818a3c8dd1bb
JOB LOG    : /home/cleber/avocado/job-results/job-2021-04-28T17.39-9e8c9e7/job.log
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 1 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2021-04-28T17.39-9e8c9e7/results.html
JOB TIME   : 4.10 s

$ avocado run --test-runner=nrunner single_fail.py 
JOB ID     : 781bf324778b2d8c0cf473f8d24cd186e538ae39
JOB LOG    : /home/cleber/avocado/job-results/job-2021-04-28T17.39-781bf32/job.log
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 1 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2021-04-28T17.39-781bf32/results.html
JOB TIME   : 4.11 s

$ avocado run --test-runner=nrunner multipletests_success.py
JOB ID     : 05cb7f6f8d01439dfb01c501290382db32fa5c38
JOB LOG    : /home/cleber/avocado/job-results/job-2021-04-28T17.39-05cb7f6/job.log
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 3 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2021-04-28T17.39-05cb7f6/results.html
JOB TIME   : 4.12 s

$ avocado run --test-runner=nrunner multipletests_fail.py 
JOB ID     : bc9bc431aeb082d02b7c31f17b79693284dabe51
JOB LOG    : /home/cleber/avocado/job-results/job-2021-04-28T17.42-bc9bc43/job.log
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 3 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2021-04-28T17.42-bc9bc43/results.html
JOB TIME   : 4.89 s
```

Which is not what's expected.  FIY, `avocado-runner-requirement-package` works as expected:

```
$ avocado-runner-requirement-package runnable-run action=check name=vim-common 
{'status': 'started', 'time': 115365.304748622}
{'status': 'running', 'time': 115365.807729953}
{'status': 'running', 'time': 115366.308372409}
{'status': 'running', 'time': 115366.808812242}
{'result': 'pass', 'returncode': 0, 'stdout': b'', 'stderr': b"dnf module for Python is required. Using the basic support from rpm and dnf commands\nRunning 'rpm -q vim-common'\nCommand 'rpm -q vim-common' finished with 0 after 0.011508703231811523s\nPackage vim-common already installed\n", 'status': 'finished', 'time': 115366.809158418}
```

File: avocado/core/runners/requirement_package.py
Comment (richtja) [inline avocado/core/runners/requirement_package.py:69]: Hi @willianrampazzo this is not compatible with the new nrunner communication from #4512. If you want to store the `stdout` and `stderr` to the logs. You have to send the `stdout` and `stderr` messages before the finish message.

File: avocado/plugins/runner_nrunner.py
Comment (willianrampazzo) [inline avocado/plugins/runner_nrunner.py:183]: @clebergnu, this was the code adjusted to do not account for non-test tasks.

**Comment by willianrampazzo** [comment]
> @willianrampazzo I understand you did a rebase, but are you sure the Task category has been accounted for here? This is the result of my experiments:
> 

Really don't know what happened. I tested locally and the results are the same as the description of this PR. Just to be on the safe side, I rebased again today and force-pushed. Let me know if it works for you.

File: avocado/core/runners/requirement_package.py
Comment (willianrampazzo) [reply avocado/core/runners/requirement_package.py:69]: Indeed, I forgot to change here. I'll update it for the next version.

**Comment by willianrampazzo** [comment]
> > @willianrampazzo I understand you did a rebase, but are you sure the Task category has been accounted for here? This is the result of my experiments:
> 
> Really don't know what happened. I tested locally and the results are the same as the description of this PR. Just to be on the safe side, I rebased again today and force-pushed. Let me know if it works for you.

@clebergnu to be on the safe side, I tested this branch on a new isolated container, and worked as expected.

**Comment by richtja** [comment]
> > > @willianrampazzo I understand you did a rebase, but are you sure the Task category has been accounted for here? This is the result of my experiments:
> > 
> > 
> > Really don't know what happened. I tested locally and the results are the same as the description of this PR. Just to be on the safe side, I rebased again today and force-pushed. Let me know if it works for you.
> 
> @clebergnu to be on the safe side, I tested this branch on a new isolated container, and worked as expected.

Hi @willianrampazzo I tried it on my machine after your force-push, and unfortunately I have the same results as @clebergnu

**Comment by willianrampazzo** [comment]
> > > > @willianrampazzo I understand you did a rebase, but are you sure the Task category has been accounted for here? This is the result of my experiments:
> > > 
> > > 
> > > Really don't know what happened. I tested locally and the results are the same as the description of this PR. Just to be on the safe side, I rebased again today and force-pushed. Let me know if it works for you.
> > 
> > 
> > @clebergnu to be on the safe side, I tested this branch on a new isolated container, and worked as expected.
> 
> Hi @willianrampazzo I tried it on my machine after your force-push, and unfortunately I have the same results as @clebergnu

The first example in the description was wrong. It had an invalid requirement, making the test to be skipped. I fixed it and should be good now. The others are correct.

**Review by clebergnu — COMMENTED** [review]
> > > @willianrampazzo I understand you did a rebase, but are you sure the Task category has been accounted for here? This is the result of my experiments:
> > 
> > 
> > Really don't know what happened. I tested locally and the results are the same as the description of this PR. Just to be on the safe side, I rebased again today and force-pushed. Let me know if it works for you.
> 
> @clebergnu to be on the safe side, I tested this branch on a new isolated container, and worked as expected.

I also tested on a VM, with the same results.  Then with further debugging, I came up with this:

```
Task "1-requirement-package-vim-common" finished message: "{'result': 'error', 'returncode': 1, 'stdout': b'', 'stderr': b"dnf module for Python is required. Using the basic support from rpm and dnf commands\nRunning '/usr/bin/sudo -n /usr/bin/dnf -y install vim-common'\nCommand '/usr/bin/sudo -n /usr/bin/dnf -y install vim-common' finished with 1 after 0.008140594989527017s\nFailed to install vim-common\n", 'status': 'finished', 'time': 182062.077523729}"
```

So even though my system already has the requirement, the check is failing because it's always attempting to perform an installation.  IMO it'd be valid to fail if I did not have the package already installed, but it should succeed if I already have the package installed.

The initial solution may be very simple: do a check before attempting an installation in the `avocado-runner-requirement-package`.

File: avocado/core/task/statemachine.py
Comment (clebergnu) [inline avocado/core/task/statemachine.py:107]: The fact that the code `return`s here, after means that the entire `triage()` gets executed again later.  This is not wrong, but I think that calling `_spawner.check_task_requirements(runtime_task)` dozens or even hundreds of times is not correct too.

An optimization is to do something like:

```diff
         # handle task dependencies
         if runtime_task.task.dependencies:
-            # check of all the dependency tasks finished
-            async with self._state_machine.lock:
-                finished_tasks = [rt_task.task for rt_task
-                                  in self._state_machine.finished]
-            if not runtime_task.task.dependencies.issubset(finished_tasks):
+            while True:
+                # check of all the dependency tasks finished
                 async with self._state_machine.lock:
-                    self._state_machine.triaging.append(runtime_task)
-                    runtime_task.status = 'WAITING DEPENDENCIES'
+                    finished_tasks = [rt_task.task for rt_task
+                                      in self._state_machine.finished]
+                    if runtime_task.task.dependencies.issubset(finished_tasks):
+                        break
+                runtime_task.status = 'WAITING DEPENDENCIES'
                 await asyncio.sleep(0.1)
-                return
 
             # dependencies finished, let's check if they finished
             # successfully, so we can move on with the parent task
```

File: avocado/plugins/runner_nrunner.py
Comment (clebergnu) [inline avocado/plugins/runner_nrunner.py:158]: Consider creating a `_get_requirement_tasks()` of sorts here to improve the readability and maintainability.

**Review by clebergnu — CHANGES_REQUESTED** [review]
Hi @willianrampazzo,

Now that we've identified the reason for the different behavior I was getting, I added a number of other points here.  Looking forwards for a v4!

Thanks!

File: avocado/core/runners/requirement_package.py
Comment (willianrampazzo) [reply avocado/core/runners/requirement_package.py:12]: I have another working version using the `SoftwareManager` class. It is much more verbose but has some more logging. The downside of using the API is that we need some mechanism to update the status repo while the backend of the `SoftwareManager` is running. I'll work on it.

File: avocado/core/runners/requirement_package.py
Comment (willianrampazzo) [reply avocado/core/runners/requirement_package.py:20]: Indeed, it was a typo.

File: avocado/core/task/statemachine.py
Comment (willianrampazzo) [reply avocado/core/task/statemachine.py:107]: I see your point, and I agree with it. I was trying to create a flag to let the code know that the spawner part was already okay when a task starts the triage again but left it away.

Unfortunately, your proposal is not working. I'll investigate why and try to develop a solution to avoid the execution of the spawner side check multiple times.

File: avocado/plugins/spawners/process.py
Comment (willianrampazzo) [reply avocado/plugins/spawners/process.py:47]: Indeed, Avocado should have some mechanism to fulfill internal requirements. As an internal requirement, I don't see the need for a user to specify it in the test itself as a test requirement, but we should have a way to check those requirements depending on the spawner the test runs on.

I could not find a way to fit the current code in the Requirements Resolver structure because the Requirements Resolver is supposed to handle test requirements and not internal requirements. I see a way to implement internal requirements and much more in the proposal I made here https://github.com/avocado-framework/avocado/discussions/4552#discussioncomment-671334.

Another downside I see with the current code is the increase of `if` statements with the increase of tags that need to be verified in the spawner's code. Again, I think it is important to check the spawner's requirements, but not as a test requirement declared in the test code. This is the main reason I opted for removing it for now. I know it still exists in the `podman` spawner and my suggestion is to remove it for now and re-introduce it in a broader dependency structure as I proposed.

**Comment by willianrampazzo** [comment]
Closing in favor of https://github.com/avocado-framework/avocado/pull/4568

**Pull request closed without merging**
